### ISSUE TITLE
Adopt standard convention of SCT

### ISSUE THREAD

**Issue by paulthomson** [p-org/PSharp#132]
It would be great if P# (and P) followed a certain standard convention of SCT work; scheduling decisions should be made immediately *before* each visible operation (send, receive, etc.).

Most SCT work models concurrent execution as a sequence of transitions/steps. Each transition represents one visible operation (e.g. send, receive) by a thread (machine/actor) followed by invisible (internal) operations performed by the same thread (which are influenced by the result of the visible operation) up until immediately before the next visible operation.

Thus, the P# SCT runtime should take control immediately *before* visible operations, not after (as I believe it does now), passing the details of the next visible operation to the scheduling strategy. This is the standard convention. This allows the runtime to inspect the next transition (whether it is enabled and what it will access) and may be helpful for certain reduction techniques (see below). The reduction where only the order of send operations matter can still be applied with this approach (see below). In fact, work on *optimal* reduction strategies also use this convention, so it should never be a limitation.

I believe the visible operations are: start (dummy operation that each machine "blocks" at before it has been "created"), create, send, receive. You may also need yield. Note that the trace of operations can be used in data race detection; create enables some start operation, send enables some receive operation; no extra information needs to be collected (as long as the operations contain sufficient information to causally link them, as they should).

Only the order of sends matters. Here is an approach to reduce the search space based on this: always execute a non-send operation (if possible; it does not matter which machine is chosen; yield may need special attention) until all enabled machines are about to execute sends; at this point, make a choice about which send operation to execute next. I believe this gives even more reduction than the current approach of making scheduling decisions *after* send operations (because the scheduler will still often choose between send operations and receive operations that become unblocked). As mentioned above, this also opens up future reduction opportunities e.g. that consider which machine a send operation is sending to. This will also ensure that other SCT scheduling algorithms can be integrated into P# (and vice-versa). 

I believe it is probably best to leave the reduction approach (e.g. always choose a non-send operation) to the individual scheduling algorithms, rather than trying to apply an approach to all scheduling algorithms. This may seem counter-intuitive, but I believe the scheduling algorithm informs the reduction technique that can be used, and the scheduling algorithm may prefer to "see" all operations, even those that seem to be redundant (e.g. dynamic partial order reduction and data race detection code will want to "see" all operations). Also, the reduction approach will probably be trivial in most cases.

It would be great to make this change before it is too late!

**Comment by akashlal** [comment]
Thank you Paul. You are correct that we currently do scheduling after a "visible operation". You make a good case for doing it before (and also while conveying information on the type of the next operation). 

In P# we don't have a yield (I suppose you want this for custom synchronization) but we have a "choice" operation (for non-deterministic choices). A choice operation commutes with all other operations but we still need to record it.

Do you have more information on "only send matters"? Why is it complete? 

In the world of randomized scheduling, do you see reduction opportunities if we adopt this convention that you're proposing? That is, how can I make testing find bugs faster? I understand the argument for systematic testing, where the total number of schedules is reduced, but what about randomized testing?

**Comment by paulthomson** [comment]
Regarding why only send matters: I believe every other operation (except maybe yield or nondet choice), when enabled, is a "partial-order operation" (or, from the state, there is a persistent set containing only this operation at the given state). In other words, if you were to synthetically disable this operation, there is no other reachable operation that conflicts with it. Thus, if you were to execute it at a later point, you could *always* commute it backwards so that it was executed immediately, without changing the state that is reached. Thus, you should execute it immediately to achieve reduction. You could essentially remove these operations from the recorded schedule entirely, but this is probably unnecessarily complicated for replay. 

For example, consider a receive operation. Once it is enabled, no other operation can change what it will do. The same applies to start and create. 

Regarding randomized scheduling: yes, I believe there are opportunities. In fact, the algorithm/approach I described in the first post (but assuming a random choice when choosing between sends) is an example of this. It has more reduction than the current random scheduler, which may end up "using a choice" to decide between send and non-send operations, which is redundantly increasing the "choice space". Similarly, when doing a PCT-like algorithm, only the send operations need to be regarded as steps; thus, the schedule length (and "choice space") is decreased. 

For some more powerful reduction ideas: a random algorithm could use the ideas of DPOR as a heuristic. E.g. detect send operations from different machines that were sent to the same machine and try reversing that order, instead of redundantly exploring different orders of *all* sends; e.g. if you have two timer machines that "ping" two different machines, there is no need to explore the different orders of the send operations from the timers machines, as they are independent.

**Comment by akashlal** [comment]
Paul,
I am not convinced. Lets chat about this offline.

Akash

**Comment by shazqadeer** [comment]
In P, there are three kinds of "receive" and two kinds of "send" operations.  The three kinds of receives are:
1. The initial "receive" executed by a state machine that starts execution right after it was created.  This operation is a right mover.
2. A receive operation that dequeues an event from the input buffer.  This operation is a right mover.
3. A receive operation that "dequeues" the null event; this happens only if there is a handler for "null" and it is not possible to dequeuer a real event from the input buffer based on the current control state.  This operation is a non mover.

The two kinds of "send" operations are:
1. An actual send operation. This operation is a non mover.  
2. A new operation. This operation is a left mover but we chose to model it as a non mover to keep the introduction of scheduling points simple.

P uses Lipton's reduction to justify its scheduling points.  For each state machine, it introduces one scheduling point at the very beginning of its execution, one scheduling point right after each "send" (i.e., send or new), one scheduling point after each receive that "dequeues" null, and one scheduling point when it blocks because it cannot dequeue any event.  This strategy guarantees that the sequence of operations in each state machine from scheduling point to the next looks like R*NL*; in fact, it looks like R*N since all left movers (essentially local operations) are right movers as well.  Notice that monitors are executed atomically with send statements but that does not affect the analysis above because send statements are non movers.

Finally, P also has a statement called "announce":
announce e, v;
This statement broadcasts event e to all monitors that observe e (similar to a send statement).  It offers extra expressiveness in case the programmers wants to write safety specifications that cannot be expressed with the events used for normal execution.  P does not add a scheduling point before or after the announce statement.  I justified this approach by saying that the announce instrumentation is being done on the reduced semantics and that the programmer is expected to be aware of this distinction.

I did not use DPOR terminology in my justification above because I find that terminology difficult to understand.  Lipton's reduction is far easier and seemed adequate for my reduction needs at the time.  

If we choose to move scheduling points before send operations, we will have to revisit the algorithm described above for inserting scheduling points.

**Comment by paulthomson** [comment]
Thanks Shaz.

Update: Pantazis and I are currently working on making these changes in P#. 

Pantazis told me about the default handler last week;  how important/used is this feature? I feel like it might remove a lot of reduction opportunity.  


More motivation below: 

I believe the approach I am suggesting is used by every other SCT tool I know of, excluding P/P# and including CHESS, if that helps. 

A visible operation may block. As such, you must (at some point) place a scheduling point immediately *before* a visible operation. There is no need to ever split up a visible operation from the invisible operations that follow it (this is proven; just execute them immediately). Thus, there is no need to place a scheduling point anywhere else. Also, putting a scheduling point at any other location means you don't know what the next visible operation is; why would you not want to have this information, since you can *provably* continue executing until you reach it? It is typically useful to know, especially for reduction techniques. It is also inconsistent, as you will have to have a scheduling point before some visible operations. Thus, each transition is: a visible operation followed by the invisible operations. A schedule is just an interleaving of these transitions; i.e., path in this transition system. 

I think the representation of transitions/operations (or, the "model of a concurrency program") can (and should) be orthogonal to the reduction strategy. Thus, the scheduling strategy should see *all* visible operations. Thus, you must place a scheduling point immediately before *every* visible operation, even if you think it will never be used. The scheduling strategy can do the reduction, or the strategy can be wrapped in another strategy that does the reduction.

**Comment by akashlal** [comment]
Is it fair to say that Paul is looking at program steps as either a non mover (visible) or a both mover (invisible)? Then the scheduling only needs to happen before a non mover.

Akash

**Comment by shazqadeer** [comment]
Sounds like it.  Putting a scheduling point before every receive and every send would also be sound, provided there is no infinite execution of invisible operations or there is a way of detecting and failing an infinite sequence of invisible operations.

In the scheme I had outlined, I was avoiding putting scheduling either before or after each successful dequeue.  It is correct though that if the dequeue blocks, a scheduling point has to be inserted.

**Comment by paulthomson** [comment]
I should clarify that scheduling only before visible (and not invisible) operations does achieve *some* reduction, but more reduction should be achieved by even the most trivial scheduling strategies. E.g. even though a scheduling point will always be placed in front of every create, start, send, receive, and default receive, most scheduling strategies will only need to consider sends and default receives. Thus, in PCT, the "number of steps" will be the number of sends and default receives; a priority change point will only every occur at these points. A delay (in delay bounding) would only ever occur after a send or default receive.

**Comment by paulthomson** [comment]
Actually, delay bounding is an exception, so ignore that. Instead: a depth first search would only consider sends and default receives.

**Comment by shazqadeer** [comment]
@paulthomson: Are you claiming that a non-default receive is a left mover, i.e., we can soundly remove scheduling points before a non-default receive?

**Comment by shazqadeer** [comment]
@paulthomson: I think you are and I am beginning to agree with you.  It looks like an alternative sound strategy for scheduling points is just before every send and every non-default receive. Of course, in addition, there is a forced scheduling point at each point where a machine blocks and at the beginning of each machine.  

Perhaps the forced scheduling point at the beginning of each machine can be avoided also but it is beginning to look complicated.

**Comment by paulthomson** [comment]
Here is the approach in more detail with some examples, as soon to be implemented in P# (hopefully).

When the P# testing runtime executes a schedule, it will repeatedly invoke the `Schedule` method at each scheduling point. The `Schedule` method is implemented by the scheduling strategy; e.g. PCT, DFS, Random, etc. The `Schedule` method receives a list of process infos, which we will show as a triple of the form: 
`(ProcessId, OperationType, Target)`. 

The operation type and target represent the next visible operation and its target. An example for a scheduling point in a multithreaded program before thread 0 is about to lock mutex `m` is: 
`(0, mutex_lock, m)`. 

The `Schedule` method returns an index specifying which machine from the list should be scheduled.

Thus, the calls to `Schedule` for a P# program that is executing from the start will be similar to the following. I will use `*` to indicate that a machine is disabled.

```
Schedule([ (m0, start,   m0)                       ]) = 0
Schedule([ (m0, receive, m0)                       ]) = 0
Schedule([ (m0, create,  m1)                       ]) = 0
Schedule([ (m0, send,    m1), (m1, start,   m1)    ]) = 1
Schedule([ (m0, send,    m1), (m1, receive, m1)*   ]) = 0
```

Note that the first two calls to schedule could be omitted; it does not really matter. The key point is that **the scheduling strategies will be notified about ALL scheduling points** and it is up to the scheduling strategies to achieve reduction.

Consider the following extract from a trace of calls to `Schedule`:

```
Schedule([ (m0, receive, m0), (m1, send, m0), (m2, receive, m2)    ]) = 0 (forced)
Schedule([ (m0, send,    m1), (m1, send, m0), (m2, receive, m2)    ]) = 2 (forced)
Schedule([ (m0, send,    m1), (m1, send, m0), (m2, receive, m2)    ]) = 2 (forced)
Schedule([ (m0, send,    m1), (m1, send, m0), (m2, send, m2)       ]) = (make a choice)
```

Most scheduling strategies will use the following rule: FORCE schedule the first enabled machine that is not about to do a `send` or `default receive`. Thus, all of the above scheduling points, except the last, are "forced". Then, the scheduling strategy will actually do whatever algorithm it wants to pick between the sends.

Note that this is better than the "decide when we can context switch" approach (i.e. don't context switch away from a machine unless we are about to execute operation type X). For example, if a machine gets blocked due to a receive, the context switch approach would now potentially have to choose between several machines that are about to execute `receive` and some that are about to execute `send`; but this is redundant; the best simple strategy is to execute all the `receive`s (`create`, `start`, etc.) and ONLY THEN make a choice. But there is no one-size fits-all approach. DPOR (or randomized DPOR-like techniques) will use different techniques to achieve even greater reduction.

**Comment by pdeligia** [comment]
Done in #198.

**Comment by paulthomson** [comment]
It turns out this is not done. We are missing scheduling points before each `Receive` operation. A `Receive` operation dequeues/receives an event and processes it. A `Receive` operation is only enabled when an event is about to be processed; a `Receive` operation should not necessarily become enabled when an event is added to the Inbox because the event may be deferred or ignored.

I think the following change will be key to make this possible. In bug finding mode, when machineA sends to machineB, machineA should execute `machineB.GetNextEvent()` (or a similar method) so that machineA knows whether to enable machineB. In production, this will not happen; instead, a task is created for machineB which may end up completing without processing any events (as is currently the case for both production and bug finding modes).

**Comment by pdeligia** [comment]
We normally do not have DEQUEUE scheduling points as these are not required: as long as you schedule at a SEND its enough to explore all possible interleavings. However, I understand that for a DPOR-based scheduler we want to introduce a DEQUEUE scheduling point. Non-DPOR schedulers can simply ignore this scheduling point (by returning the same scheduling choice as the previous one).

I am not very sure about the last part, we should discuss this.

**Comment by paulthomson** [comment]
Notes:

Add scheduling point after call to GetNextEvent (possibly in NotifyDequeuedEvent) but before handling the event.

Handling a Raise should not be a scheduling point. (Already OK)
The initial event should not cause a Receive scheduling point. (Already OK)

Problem 1: When an event handler exits, it will hit a `Receive` scheduling point. When the event handler is recreated, it must not hit another `Receive` scheduling point before processing an event.
Solution: Flag to skip receive if the last receive was exiting event handler (a receive where we are disabled).

Problem 2: Only enable a machine when it can actually handle an event.
Solution: Change GetNextEvent so that it can be called from the sender machine. It will now return an index to the next event (without removing it). It will also put a raised event or a default event in the Inbox if needed. The sender machine will now know whether to enable the receiving machine.

**Comment by pdeligia** [comment]
I am not sure if I want to have all these guys be a `Receive` scheduling point. In the above discussion there are (at least) the following `Receive` points:
- `Receive` when an `ISchedulable` dequeues an event from its inbox.
- `Receive` when the event handler of an `ISchedulable` exits because the inbox is empty (or there is nothing that can be dequeued, e.g. all remaining events are deferred).
- `Receive` when an `ISchedulable` asynchronously "blocks" its handler until it receives an event its waiting for.

I understand that in non-P-based SCT frameworks all the above receives might be treated the same (well mostly because the third one does not exist), however I don't think this is easy in our case. The problem is that regular schedulers in P# need to differentiate between all the above receive points.

For example, I don't want the naïve random scheduler to schedule at a dequeue, because this is not required (scheduling at all sends is enough to expose all bugs). However, a regular scheduler still needs to schedule at the end of a handler or at a receive statement, else the program will deadlock. For this reason, I want to add an extra operation called `Wait`, and the last two `Receive` points above, are essentially a `Wait`. I think this is a useful optimization in general (even for third-party actor-based SCT frameworks), since sends and receives commute -- and most schedulers can safely ignore receives.

A `Wait` operation denotes that you are waiting for some arbitrary trigger to happen (scheduler does not need to know, can be agnostic). DPOR can process `Receive` though to perform partial-order reduction, and can essentially treat `Receive` and `Wait` the same.

Another option would be to treat `Receive` differently depending on the machine being enabled or not (in dequeue the machine is enabled, in the last 2 points its disabled), however, I quite like to know about this just by querying the operation, rather than having to check both the operation and the enabled status of the machine. But I am happy to debate about this.

Btw, beyond what has been discussed, there is a potential problem because an extra scheduling point can currently trigger the liveness checker to assert liveness / capture a state, making it inefficient (and potentially buggy with the current implementation). Need to be careful there because its very easy to introduce bugs. I know that the liveness checker could be implemented as a strategy, which could make solving these problems much easier, but that requires some heavy refactoring which we don't have the time to do right now. So we might want to initially disable the liveness checker for DPOR (and receive operations) until this is investigated/resolved properly.

@paulthomson we can discuss :-)

**Comment by pdeligia** [comment]
Problem 1 above solved. @paulthomson we can code review together.

**Comment by paulthomson** [comment]
I think you are saying you want to add a new OperationType, `Wait`, and some of the receive scheduling points will use this new operation type. I see no downside to this, except the second receive operation type must remain as a receive (see below for an example). Also note that I think the third receive operation type (a receive statement) should, in some cases, not block at all (mentioned below---the required event may already be in the Inbox, right?).

> The problem is that regular schedulers in P# need to differentiate between all the above receive points.

I don't see why, but it doesn't hurt to differentiate the async receive statement (the third case).

> For example, I don't want the naïve random scheduler to schedule at a dequeue, because this is not required

Indeed, the random scheduler does not need to (as explained in my example in the first comment), but I don't know why differentiating is needed. In fact, "not scheduling at a dequeue" is not enough; more can be done, as I explained in my second comment.

> However, a regular scheduler still needs to schedule at the end of a handler or at a receive statement, else the program will deadlock.

Essentially: if a machine blocks, another machine must be scheduled. I agree. But I still don't see why differentiating is needed.

> ...the last two Receive points above, are essentially a Wait

I think the second one really must be a receive, not a wait, because the wait will sometimes replace a receive (because they are actually the same operation!). Consider the following schedule:

```
M1      M2

send    
send    
        receive
        receive
```

M1's first send is received by M2's first receive, and M1's second send is received by M2's second receive. M1's second send and M2's first receive are independent, so can be commuted without changing the state reached. Let's look at this schedule now:

```
M1      M2

send    
        receive
send    
        wait
```

M2's second receive has changed to a wait! Thus, these schedules are no longer equivalent. Thus, the use of wait is incorrect. 

> I think this is a useful optimization in general

I see no optimization (indeed, the current approach of not scheduling before a dequeue is nonoptimal), but again I think it is fine to differentiate the async receive statement, if you want.

> since sends and receives commute

I don't agree with this in general.

> and most schedulers can safely ignore receives

They can do this with a scheduling point before every receive and without differentiating the different receive types; they should force execution of all enabled receive operations before making a choice. In fact, this is more efficient than the current approach of not scheduling before a dequeue. 

> A Wait operation denotes that you are waiting for some arbitrary trigger to happen (scheduler does not need to know, can be agnostic).

A receive also denotes this. The DPOR scheduler *does* need to know which send eventually enables the operation, regardless of whether the operation was initially disabled.

> Another option would be to treat Receive differently depending on the machine being enabled or not (in dequeue the machine is enabled, in the last 2 points its disabled), however, I quite like to know about this just by querying the operation, rather than having to check both the operation and the enabled status of the machine. But I am happy to debate about this.

Actually, a receive statement could also be initially enabled if the required event already happens to be in the Inbox, right? Thus, I think you would still need to differentiate the receive statement (although I still don't know why this would be useful). As I have argued above, the second receive operation really is a receive operation.

> Btw, beyond what has been discussed, there is a potential problem because an extra scheduling point can currently trigger the liveness checker to assert liveness / capture a state, making it inefficient. ...So we might want to initially disable the liveness checker for ... receive operations ...

I think that is a good solution and so there is no problem. By the way, I think you should also disable it for `create`, `start`, `halt`, as well, which I assume you don't currently do (and so I hope this can motivate further why efficiency/reduction vs. correctly having a scheduling point before each visible operation are orthogonal). In other words, you should only consider `send`, as I mentioned in the first comment of this issue. There are exceptions and it depends on the scheduler. E.g. preemption bounding and delay bounding (which are not compatible with the live checker anyway, right?) must consider `create`; a machine that infinitely creates machines without issuing any sends would be an issue, but you could detect this.

For me, the takeaway is that these are all receive operations: a receive operation is enabled by a send operation, and the send operation may or may not have already occurred. The async receive statement is exactly the same (it may or may not be blocked initially, I hope), but it is fine to give it a different name.

**Comment by pdeligia** [comment]
Lets discuss in person as its much easier. I think I can understand clearer what you have in mind now (and I start to agree with some of your points).

Just to quickly say:
- If a machine uses the receive statement, and the event waiting to receive is already in the inbox, then the scheduling point is not inserted there currently. I believe that is fine for our existing schedulers (similar to how we do not schedule at a dequeue), but I just realised that we need to add it for DPOR to work, right? Your comments regarding `IsEnabled` above are correct!

In general, this part of the code is really complicated, this is why I want to be 100% convinced before I make any big changes. I don't disagree with your points (and I think I start getting convinced why we only need `Receive` and not `Wait`, although this requires some refactoring and doing what you said with forcing the `Receive` operations execute in current schedulers), but we need to discuss more so that I can fully understand this.

I think we should work on two branches: (1) where we try to convert the liveness checker into a strategy which can sit on top (or below) any other fair strategy, and (2) the changes you mention in this thread that are needed by DPOR so that I can also better understand them. (2) will most likely conflict with existing liveness implementation until (1) is done, so I don't want to merge them to master before (1). If we are satisfied with all this exercise, we can then merge (1) and then (2). But meanwhile, we can do DPOR experiments in (2), without liveness initially (which is though really important for the applications we are targeting, so we need to support it).

We should try to also discuss with @akashlal when we meetup on Tuesday.

**Comment by paulthomson** [comment]
Yeah in person is easier!

> If a machine uses the receive statement, and the event waiting to receive is already in the inbox, then the scheduling point is not inserted there currently. ..  I just realised that we need to add it for DPOR to work, right?

A receive statement is indeed a visible operation and so a scheduling point is required. I think a nice argument that I probably should have come up with much earlier is that:

* A visible operation is an operation that changes the shared state (including e.g. the state of an Inbox) and/or can become enabled/disabled and/or can enable/disable another operation.
* Visible operations *are* essentially scheduling points.
* Where you put scheduling points can (and should) be orthogonal to any optimisation/reduction. This is especially important since some reduction techniques only work with certain schedulers.

Thus, I think of this change as not simply being necessary for DPOR, but instead being consistent with the rest of the scheduling points and with other SCT work. 

Perhaps the liveness checker could continue to work while we add scheduling points, as long as we carefully update when the liveness checker is triggered. But your approach sounds good too.

**Comment by paulthomson** [comment]
I just want to document a change I have made in the `dpor3` branch (in case this branch gets lost).

The scheduling point for send should be inserted *before* the check for whether the send will "fail" because the target machine has halted. One way to reason about this is that whether the send fails or not depends on when we schedule the send, so we should have a scheduling point to ensure we explore both possible outcomes. 

Another way of looking at it is that a send to a machine should potentially be independent with the final operation of the target machine. E.g.:

```
m0					m1
					send (to m0)
send (to m1)    	
[halt]
```

The two sends should be independent and so can be swapped without changing the state reached (assuming conceptually that, when halted, m0 will clear its mailbox and not add further messages to its mailbox, or that m0 does not clear its mailbox and does add all future messages to its mailbox). If we try to execute this schedule:

```
m0					m1
send (not to m1)    	
[halt]
					[send has gone]
```

...m1's send has now vanished and so we would have to treat these sends as dependent, even though they aren't really.

**Comment by pdeligia** [comment]
I think what you say makes sense. Btw, I just merged the liveness strategy refactoring, so we might want to rebase master to dpor branch!

**Comment by paulthomson** [comment]
Thanks! I have rebased!

I am bit confused as to what `PrepareForNextChoice` is for. Couldn't all that code just go in `GetNext`? They are called one after the other in `BugFindingScheduler.Schedule`, so it seems like `PrepareForNextChoice` serves no purpose. It sounds like you want to do something just before the next scheduling choice, which is precisely what `GetNext` provides you, right?

**Comment by pdeligia** [comment]
Scheduling strategies have now been separated in `Microsoft.TestingServices.SchedulingStrategies.dll` [here](https://github.com/p-org/PSharp/tree/master/Source/AddOns/SchedulingStrategies). We can rebase dpor3 now on top of this. We should also add an `IContract` (i.e. asserter interface) similar to `ILogger` ([here](https://github.com/p-org/PSharp/blob/master/Source/AddOns/SchedulingStrategies/Utilities/ILogger.cs)) and `IRandomNumberGenerator` ([here](https://github.com/p-org/PSharp/blob/master/Source/AddOns/SchedulingStrategies/RandomNumberGenerators/IRandomNumberGenerator.cs)) that can assert the implementation of scheduling strategies (only used in DPOR right now).

I removed the `PrepareForNextChoice`!

**Comment by pdeligia** [comment]
@paulthomson btw, I am not 100% sure about the name of the new dll, but I wanted to emphasize that these scheduling strategies are specifically for controlled/systematic testing, that's why I decided to not just use `Microsoft.SchedulingStrategies.dll` :-)

Another idea would be to rename it to `Microsoft.PSharp.TestingServices.SchedulingStrategies.dll`. Because, `TestingServices` can mean many different things ... to emphasize its about P#-style SCT (other SCT tools shouldn't really care).

If you think of any better ideas please say of course!

**Comment by pdeligia** [comment]
@paulthomson I was reviewing the code and I am a bit confused why we are using a scheduling point inside `NotifyDefaultEventHandlerCheck`. What if the default handler never triggers? We have just done an extra scheduler, unless I am missing something (which could be the case)!

I wonder if we can remove default all together ... not sure if this is even used really. Lets chat with @akashlal next time.

**Comment by paulthomson** [comment]
See #206. If the machine does not have a default state handler or the machine never gets into a state with a default state handler, there will be no extra scheduling point. Otherwise, I feel it is (unfortunately) needed. But if you don't use the feature, then there is no penalty anyway.

**Comment by pdeligia** [comment]
Cool, I will check it out and let you know! Btw, the rest of the changes look awesome :-) I just did some refactoring and rebased (primarily all DPOR stuff are now internal and in their own namespace to keep things more tidy).

Need to fix the failing tests now (3 integration need to get new seeds, and the one unit test not sure what is happening, might have to do that we have extra steps now so need to change test parameters, will investigate).

Then we also need to take some measurements with random+pct+fairpct using the extra scheduling points, and also the other strategy that we discussed (all operations forced, until only sends).
